webapi/Controllers/ChatController.cs

```diff
@@ -194,87 +196,112 @@ private async Task RegisterFunctionsAsync(Kernel kernel, Dictionary<string, stri
         }
 
         if (variables.TryGetValue("customPlugins", out object? customPluginsString))
         {
-            CustomPlugin[]? customPlugins = JsonSerializer.Deserialize<CustomPlugin[]>(customPluginsString!.ToString()!);
+            tasks.Add(this.RegisterCustomPlugins(kernel, customPluginsString, authHeaders));
+        }
+
+        await Task.WhenAll(tasks);
+    }
+
+    private async Task RegisterGithubPlugin(Kernel kernel, string GithubAuthHeader)
+    {
+        this._logger.LogInformation("Enabling GitHub plugin.");
+        BearerAuthenticationProvider authenticationProvider = new(() => Task.FromResult(GithubAuthHeader));
+        await kernel.ImportPluginFromOpenApiAsync(
+            pluginName: "GitHubPlugin",
+            filePath: Path.Combine(this._pluginDirectoryPath, "GitHubPlugin/openapi.json"),
+            new OpenApiFunctionExecutionParameters
+            {
+                AuthCallback = authenticationProvider.AuthenticateRequestAsync,
+            });
+    }
+
+    private async Task RegisterJiraPlugin(Kernel kernel, string JiraAuthHeader, KernelArguments variables)
+    {
+        this._logger.LogInformation("Registering Jira plugin");
+        var authenticationProvider = new BasicAuthenticationProvider(() => { return Task.FromResult(JiraAuthHeader); });
+        var hasServerUrlOverride = variables.TryGetValue("jira-server-url", out object? serverUrlOverride);
+
+        await kernel.ImportPluginFromOpenApiAsync(
+            pluginName: "JiraPlugin",
+            filePath: Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "Plugins", "OpenApi/JiraPlugin/openapi.json"),
+            new OpenApiFunctionExecutionParameters
+            {
+                AuthCallback = authenticationProvider.AuthenticateRequestAsync,
+                ServerUrlOverride = hasServerUrlOverride ? new Uri(serverUrlOverride!.ToString()!) : null,
+            });
+    }
 
-            if (customPlugins != null)
+    private Task RegisterMicrosoftGraphPlugins(Kernel kernel, string GraphAuthHeader)
+    {
+        this._logger.LogInformation("Enabling Microsoft Graph plugin(s).");
+        BearerAuthenticationProvider authenticationProvider = new(() => Task.FromResult(GraphAuthHeader));
+        GraphServiceClient graphServiceClient = this.CreateGraphServiceClient(authenticationProvider.GraphClientAuthenticateRequestAsync);
+
+        kernel.ImportPluginFromObject(new TaskListPlugin(new MicrosoftToDoConnector(graphServiceClient)), "todo");
+        kernel.ImportPluginFromObject(new CalendarPlugin(new OutlookCalendarConnector(graphServiceClient)), "calendar");
+        kernel.ImportPluginFromObject(new EmailPlugin(new OutlookMailConnector(graphServiceClient)), "email");
+        return Task.CompletedTask;
+    }
+
+    private async Task RegisterCustomPlugins(Kernel kernel, object? customPluginsString, Dictionary<string, string> authHeaders)
```

Comment:
Since we might be registering more than one custom plugin, it seems to me that it would make more sense to have this method return a IEnumerable of Tasks (perhaps a List) and then do tasks.AddRange() so that these custom plugins can be loaded in parallel too.

For example:

tasks.AddRange(RegisterCustomPlugins(...));

Reply:
RegisterCustomPlugins is now returning IEnumerable<Task> and yield returning the result of kernel.ImportPluginFromOpenAIAsync